db/encoding/bytes.go

```diff
@@ -0,0 +1,192 @@
+// different terminators for things such as JSON key encoding.
+func encodeBytesAscendingWithTerminatorAndPrefix(
+	b []byte, data []byte, terminator byte, prefix byte,
+) []byte {
+	b = append(b, prefix)
+	return encodeBytesAscendingWithTerminator(b, data, terminator)
+}
+
+// encodeBytesAscendingWithTerminator encodes the []byte value using an escape-based
+// encoding. The encoded value is terminated with the sequence
+// "\x00\terminator". The encoded bytes are append to the supplied buffer
+// and the resulting buffer is returned. The terminator allows us to pass
+// different terminators for things such as JSON key encoding.
+func encodeBytesAscendingWithTerminator(b []byte, data []byte, terminator byte) []byte {
+	bs := encodeBytesAscendingWithoutTerminatorOrPrefix(b, data)
+	return append(bs, escape, terminator)
+}
+
+// encodeBytesAscendingWithoutTerminatorOrPrefix encodes the []byte value using an escape-based
+// encoding.
+func encodeBytesAscendingWithoutTerminatorOrPrefix(b []byte, data []byte) []byte {
+	for {
+		// IndexByte is implemented by the go runtime in assembly and is
+		// much faster than looping over the bytes in the slice.
+		i := bytes.IndexByte(data, escape)
+		if i == -1 {
+			break
+		}
+		b = append(b, data[:i]...)
+		b = append(b, escape, escaped00)
+		data = data[i+1:]
+	}
+	return append(b, data...)
+}
+
+// EncodeBytesDescending encodes the []byte value using an
+// escape-based encoding and then inverts (ones complement) the result
+// so that it sorts in reverse order, from larger to smaller
+// lexicographically.
+func EncodeBytesDescending(b []byte, data []byte) []byte {
+	n := len(b)
+	b = EncodeBytesAscending(b, data)
+	b[n] = bytesDescMarker
+	onesComplement(b[n+1:])
+	return b
+}
+
+// DecodeBytesAscending decodes a []byte value from the input buffer
+// which was encoded using EncodeBytesAscending. The decoded bytes
+// are appended to r. The remainder of the input buffer and the
+// decoded []byte are returned.
+func DecodeBytesAscending(b []byte, r []byte) ([]byte, []byte, error) {
+	return decodeBytesInternal(b, r, ascendingBytesEscapes, true /* expectMarker */, false /* deepCopy */)
+}
+
+// DecodeBytesAscendingDeepCopy is the same as DecodeBytesAscending, but the
+// decoded []byte will never alias memory of b.
+func DecodeBytesAscendingDeepCopy(b []byte, r []byte) ([]byte, []byte, error) {
+	return decodeBytesInternal(b, r, ascendingBytesEscapes, true /* expectMarker */, true /* deepCopy */)
+}
+
+// DecodeBytesDescending decodes a []byte value from the input buffer
+// which was encoded using EncodeBytesDescending. The decoded bytes
+// are appended to r. The remainder of the input buffer and the
+// decoded []byte are returned.
+//
+// Note that this method internally will always perform a deep copy, so there is
+// no need to introduce DecodeBytesDescendingDeepCopy to mirror
+// DecodeBytesAscendingDeepCopy.
+func DecodeBytesDescending(b []byte, r []byte) ([]byte, []byte, error) {
```

Comment:
todo: `r` appears to always be empty?

Reply:
it might not be used for the current use-case but it's quite handy if you want to decode bytes one after another into a preallocated buffer:
```go
r := make([]byte, 0, bigBuffer)
b, r, err = DecodeBytesAscending(b, r)
// check err
b, r, err = DecodeBytesAscending(b, r)
...
```

Reply:
Most code is handy in some situations, I don't like reading and maintaining code that is not useful to our situations.

Reply:
done